pkg/plugins/policies/jsonpatch/validators/validator.go

```diff
@@ -0,0 +1,112 @@
+package validators
+
+import (
+	"encoding/json"
+
+	"github.com/kumahq/kuma/api/common/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/validators"
+)
+
+func ValidateJsonPatchBlock(
+	rootPath validators.PathBuilder,
+	patchBlock []v1alpha1.JsonPatchBlock,
+) validators.ValidationError {
+	var err validators.ValidationError
+
+	if len(patchBlock) == 0 {
+		err.AddViolationAt(rootPath, validators.MustNotBeEmpty)
+	}
+
+	for i, patch := range patchBlock {
+		path := rootPath.Index(i)
+
+		err.AddErrorAt(path.Field("op"), validateOp(patch.Op))
+		err.AddErrorAt(path.Field("from"), validateFrom(patch.From, patch.Op))
+		err.AddErrorAt(path.Field("value"), validateValue(patch.Value, patch.Op))
+		err.AddErrorAt(path.Field("path"), validatePath(patch.Path, patch.Op))
+	}
+
+	return err
+}
+
+// validateOf checks if "op" field is valid json patch operation
+func validateOp(op string) validators.ValidationError {
+	var err validators.ValidationError
+
+	switch op {
+	case "add", "remove", "replace", "move", "copy":
+		break
+	default:
+		err.Add(validators.MakeFieldMustBeOneOfErr("op",
+			"add", "remove", "replace", "move", "copy",
+		))
+	}
+
+	return err
+}
+
+// validateFrom checks if "op" field is valid ("move" or "copy") when "from"
+// field is provided
+func validateFrom(from *string, op string) validators.ValidationError {
+	var err validators.ValidationError
+
+	switch op {
+	case "move", "copy":
+		if from == nil {
+			err.Add(validators.MakeOneOfErr("from",
+				"op",
+				validators.MustNotBeEmpty,
+				[]string{"move", "copy"},
+			))
+		}
+	default:
+		if from != nil {
+			err.Add(validators.MakeOneOfErr("from",
+				"op",
+				"is allowed only",
+				[]string{"move", "copy"},
+			))
+		}
+	}
+
+	return err
+}
+
+// validateValue checks if "op" field is valid ("add" or "replace") when "value"
+// field is provided
+func validateValue(value json.RawMessage, op string) validators.ValidationError {
+	var err validators.ValidationError
+
+	switch op {
+	case "add", "replace":
+		if value == nil {
+			err.Add(validators.MakeOneOfErr("value",
+				"op",
+				validators.MustNotBeEmpty,
+				[]string{"add", "replace"},
+			))
+		}
+	default:
+		if value != nil {
+			err.Add(validators.MakeOneOfErr("value",
```

Comment:
Not quite sure it's worth putting these two different errors into one function, where either fieldB == nil or fieldB != nil. This line is hard to read IMO. Could be the naming, dunno

Reply:
I have refactored these functions